src/centrality/truss.cpp

```diff
@@ -0,0 +1,271 @@
+#include "igraph_interface.h"
+#include "igraph_motifs.h"
+#include "igraph_community.h"
+
+#include <iostream>
+#include <vector>
+#include <unordered_set>
+
+using namespace std;
+
+/* helper functions */
+void igraph_truss_i_unpack(const igraph_vector_int_t *tri, igraph_vector_t *unpacked_tri);
+void igraph_truss_i_compute_support(const igraph_vector_t *eid, igraph_vector_int_t *support);
+int igraph_i_trussness(const igraph_t *graph, igraph_vector_int_t *support,
+  igraph_vector_int_t *truss);
+
+/**
+ * \function trussness
+ * \brief Find the "trussness" for every edge in the graph. This
+ * indicates the highest k-truss that the edge occurs in.
+ *
+ * \param graph The input graph.
+ * \param truss Pointer to initialized vector of truss values that will
+ * indicate the highest k-truss each edge occurs in. It will be resized as
+ * needed.
+ *
+ * \example examples/simple/igraph_truss.cpp
+ */
+int igraph_truss(const igraph_t* graph, igraph_vector_int_t* truss){
+  igraph_vector_int_t triangles, support;
+  igraph_vector_t eid, unpacked_triangles;
+
+  /* Manage the stack to make it memory safe */
+  IGRAPH_VECTOR_INT_INIT_FINALLY(&support, igraph_ecount(graph));
+  IGRAPH_VECTOR_INIT_FINALLY(&eid, 0);
+  IGRAPH_VECTOR_INIT_FINALLY(&unpacked_triangles, 0);
+  IGRAPH_VECTOR_INT_INIT_FINALLY(&triangles, 0);  // will be resized
+
+  // List the triangles as vertex triplets.
+  IGRAPH_CHECK(igraph_list_triangles(graph, &triangles));
+
+  // Unpack the triangles from vertex list to edge list.
+  long int num_triangles = igraph_vector_int_size(&triangles);
+  IGRAPH_CHECK(igraph_vector_resize(&unpacked_triangles, 2 * num_triangles));
+  igraph_truss_i_unpack(&triangles, &unpacked_triangles);
+  igraph_vector_int_destroy(&triangles);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Get the edge ids of the unpacked triangles. Note: a given eid can occur
+  // multiple times in this list if it is in multiple triangles.
+  IGRAPH_CHECK(igraph_vector_resize(&eid, num_triangles));
+  IGRAPH_CHECK(igraph_get_eids(graph, &eid, &unpacked_triangles, 0, 0, 1));
+  igraph_vector_destroy(&unpacked_triangles);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the support of the edges.
+  igraph_truss_i_compute_support(&eid, &support);
+  igraph_vector_destroy(&eid);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the truss of the edges.
+  IGRAPH_CHECK(igraph_vector_int_init(truss, igraph_ecount(graph)));
+  IGRAPH_CHECK(igraph_i_trussness(graph, &support, truss));
+  igraph_vector_int_destroy(&support);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  return IGRAPH_SUCCESS;
+}
+
+// Unpack the triangles as a vector of vertices to be a vector of edges.
+// So, instead of the triangle specified as vertices [1, 2, 3], return the
+// edges as [1, 2, 1, 3, 2, 3] so that the support can be computed.
+void igraph_truss_i_unpack(const igraph_vector_int_t *tri, igraph_vector_t *unpacked_tri) {
+  long int i, j;
+  for (i = 0, j = 0; i < igraph_vector_int_size(tri); i += 3, j += 6){
```

Comment:
```suggestion
  long int i, j, n = igraph_vector_int_size(tri);
  for (i = 0, j = 0; i < n; i += 3, j += 6) {
```

Because there is a cost to non-inlineable function calls in a tight-loop.